### PR TITLE
Bender integration, sram update

### PR DESCRIPTION
**Pull request by micprog**
Added `Bender.yml` file and bump dependencies for compatibility to bender.
Updated `tech_cells_generic`, changed deprecated `generic_memory` to `tc_sram`.

@bluewww @meggiman Please let me know if any other changes are required
Replaces https://github.com/pulp-platform/pulp_soc/pull/33

### CODE REVIEW

File: Bender.yml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  axi:                    { git: "git@github.com:pulp-platform/axi.git", version: 0.24.1 }
+  # axi_node:               { git: "git@github.com:pulp-platform/axi_node.git", version: 1.1.4 } # deprecated, replaced by axi_xbar (in axi repo)
+  axi_slice:              { git: "git@github.com:pulp-platform/axi_slice.git", version: 1.1.4 } # deprecated, replaced by axi_cut (in axi repo)
+  axi_slice_dc:           { git: "git@github.com:pulp-platform/axi_slice_dc.git", version: 1.1.3 } # deprecated, replaced by axi_cdc (in axi repo)
+  timer_unit:             { git: "git@github.com:pulp-platform/timer_unit.git", version: 1.0.2 }
+  common_cells:           { git: "git@github.com:pulp-platform/common_cells.git", version: 1.20.0 }
+  cluster_interconnect:   { git: "git@github.com:pulp-platform/cluster_interconnect.git", version: 1.1.1 }
+  fpnew:                  { git: "git@github.com:pulp-platform/fpnew.git", version: 0.6.5 }
+  jtag_pulp:              { git: "git@github.com:pulp-platform/jtag_pulp.git", rev: "v0.1" }
+  riscv:                  { git: "git@github.com:micprog/cv32e40p.git", rev: "pulpissimo-v3.4.1"} # To be updated to openhwgroup repository
```
Comment (meggiman) [inline Bender.yml:33]: Any update on benderization of  cv32e40p? I would prefer to avoid letting the master branch of pulp_soc to point to a user repo.

File: Bender.yml [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  udma_core:              { git: "git@github.com:pulp-platform/udma_core.git", version: 1.0.2 }
+  udma_uart:              { git: "git@github.com:pulp-platform/udma_uart.git", version: 1.0.1 }
+  udma_i2c:               { git: "git@github.com:pulp-platform/udma_i2c.git", version: 1.0.0 }
+  udma_i2s:               { git: "git@github.com:pulp-platform/udma_i2s.git", version: 1.1.1 }
+  udma_qspi:              { git: "git@github.com:pulp-platform/udma_qspi.git", version: 1.0.3 }
+  udma_sdio:              { git: "git@github.com:pulp-platform/udma_sdio.git", version: 1.1.0 }
+  udma_camera:            { git: "git@github.com:pulp-platform/udma_camera.git", version: 1.1.1 }
+  udma_filter:            { git: "git@github.com:pulp-platform/udma_filter.git", version: 1.0.2 }
+  udma_external_per:      { git: "git@github.com:pulp-platform/udma_external_per.git", version: 1.0.3 }
+  hwpe-mac-engine:        { git: "git@github.com:micprog/hwpe-mac-engine.git", rev: "v1.2-bender" } # To be updated with pulp-platform dependency
```
Comment (meggiman) [inline Bender.yml:47]: And this one? We should close open PR (where they are under our control) before merging to master...

File: src_files.yml [hunk 3]
```diff
@@ -55,7 +55,6 @@ components:
   ]
   files: [
     rtl/components/apb_clkdiv.sv,
-    rtl/components/apb_timer_unit.sv,
```
Comment (meggiman) [inline src_files.yml:58]: Why did you remove this file?

**Review by meggiman — COMMENTED** [review]
First of all, thanks a lot for the PR. I really like the fact that pulp_soc is getting benderized and also the getting rid of the legacy SRAM models is a good improvement. However, there are still a couple of issues:

1. There should not be any references to your personal git repo anymore once we merge this into master. 
2. The other problem is, that the Bender.yml and ips_list.yml seem to not be in sync. At least for RI5CY bender.yml and ips_list.yml point to different repositories. Maybe there are more inconsistencies but I haven't checked yet. In anycase, all IP versions in Bender.yml and ips_list.yml need to be aligned before merging.

File: Bender.yml (hunk 1 above)
Comment (bluewww) [reply Bender.yml:33]: I guess ri5cy needs to be patched. We now have a fork in `pulp-platform` to maintain such patches. I suggest  the patch to go there.

Comment (micprog) [reply Bender.yml:33]: The changes to the Bender.yml file in ri5cy/cv32e40p are very limited, however updating this in the cv32e40p master would not improve the situation, as updating this in pulp may have a larger effect (core renaming, ...). Ibex uses a branch in the lowrisc repo to maintain pulp modifications, a fork in pulp-platform may be an option, I believe there is an issue open for this: https://github.com/pulp-platform/pulp_soc/issues/44

Comment (bluewww) [reply Bender.yml:33]: I mean like this https://github.com/pulp-platform/cv32e40p/tree/pulpissimo-v3.4.0-rev1

File: src_files.yml (hunk 3 above)
Comment (micprog) [reply src_files.yml:58]: This file has a nearly identical version in the [timer_unit](https://github.com/pulp-platform/timer_unit) dependency. I am unsure which version is newer, and to avoid severe issues I did not delete the file yet, but duplicates can lead to other problems.

File: Bender.yml (hunk 1 above)
Comment (micprog) [reply Bender.yml:33]: Looks good, I will make a pull request to that repo with my changes to the ri5cy/cv32e40p Bender.yml file and update the dependency here once merged and tagged.

File: Bender.yml (hunk 2 above)
Comment (micprog) [reply Bender.yml:47]: The HWPE repositories have significant updates within, the versions referenced in pulp are 2-3 years old if I'm not mistaken. I believe @FrancescoConti is working on updating this, I only added a `Bender.yml` file to the versions indicated by pulp. I did not investigate the current master and the impact the changes to these dependencies may have, but assumed that it may break compatibility. A temporary solution could be to add a branch to this repository as well, but [hwpe-mac-engine](https://github.com/pulp-platform/hwpe-mac-engine) also depends on [hwpe-ctrl](https://github.com/pulp-platform/hwpe-ctrl) and [hwpe-stream](https://github.com/pulp-platform/hwpe-stream), so these would require similar treatment.

**Comment by micprog** [comment]
Small update: in my most recent tests building pulpissimo with this `pulp_soc`, I encountered an issue with two different definitions of `FLL_BUS` due to updating `apb_fll_if`, now fixed with the latest commit.

**Comment by meggiman** [comment]
The commit id of Ibex in this PR seems to be invalid or the commit doesn't exist anymore. Could you please update this?

**Comment by meggiman** [comment]
Also I just noticed that there are many version conflicts when trying a bender checkout. If possible we should have the master branch in a state, where version conflicts do not have to be resolved manually or with an override file. This will probably require a lot of version bumping to align the repositories version requirements...

**Comment by micprog** [comment]
> The commit id of Ibex in this PR seems to be invalid or the commit doesn't exist anymore. Could you please update this?

The pulpissimo branch in ibex has been moved to a newer version of ibex, referenced in #47. There are a few options to fix this, depends on what you would prefer.
1. I can reference my personal repo for this commit until #47 is merged, which still has the commit (I could add a `pulpissimo-v6.0.1` tag). 
2. Alternatively, ips_list.yml could reference the `pulpissimo-v6.0.0`, which is the version previously used, however does not include updates for bender and minor bugfixes. 
3. If the older version referenced here should originate from the lowRISC repository, the appropriate commits would need to be re-introduced, with tags (@vogelpi's help here would be needed).
4.  I could re-arrange the update to ibex in #47 such that it can be merged first, and an available version of ibex referenced in this PR.

**Comment by meggiman** [comment]
I personally would prefer option 4. Merging this PR should yield a working Bender flow which would not work with option 2.

**Comment by micprog** [comment]
> Also I just noticed that there are many version conflicts when trying a bender checkout. If possible we should have the master branch in a state, where version conflicts do not have to be resolved manually or with an override file. This will probably require a lot of version bumping to align the repositories version requirements...

While behavior would definitely be improved, these updates are not always that easy. For specifics:
- [ ] `tech_cells_generic`: requires update of `udma_sdio`, `udma_i2c`, `udma_camera`, `udma_qspi`, `cv32e40p`, `common_cells`, `apb_adv_timer` (if identical versions between ips_list.yml and bender.yml are desired, a new tag would also be needed)
- [ ] `common_verification`: requires update of `common_cells`
- [ ] `axi`: requires update of `register_interface`, and likely the removal of `axi_slice_dc`

I will start with these updates, but strongly suggest the use of a Bender.local file for overrides in the main pulpissimo and pulp repositories, as certain mismatches seem to be prevalent in pulp.

**Comment by micprog** [comment]
> I personally would prefer option 4. Merging this PR should yield a working Bender flow which would not work with option 2.

Done, I will update the dependencies in this PR once #47 is merged

**Comment by vogelpi** [comment]
Thanks for reworking #47 @micprog this is the cleanest solution.

**Comment by bluewww** [comment]
#47 is merged

**Comment by bluewww** [comment]
Can you also add the override file to this pr you use to the bender checkout? This would allow users to get a working state of this project withouth having to tell bender which ip version to use. I believe this is a good short term solution.

**Comment by meggiman** [comment]
@micprog Regarding the version alignment issues of all the dependencies I can see the dificulty, although using an override file kind of defeats the purpose of Benders semantic versioning resolution. In order to finally bring bender support to pulp_soc without further delays I would thus suggest that we temporarily add a bender.local file with the necessary overrides and later on remove it step by step once the other IPs become aligned. Otherwise requiring less experienced pulp user to decide which version to check out is a fool proof way to completely overwhelm ourselves with support requests once we point to the newest pulp_soc version in PULPissimo and pulp-open. What do you think?

**Comment by micprog** [comment]
@bluewww @meggiman Based on some quick tests, the bender config file used for overrides can currently only be placed in the top-level repository, i.e. the PULPissimo or pulp-open repositories, not this repo. This would also make sense to keep all overrides in one central location. I agree that users should not be required to choose dependency versions, unless they are working on dependency updates, so integrating Bender to the top repositories will require adding a config file to those PRs, but unfortunately adding one here will not make a difference. Is this an acceptable approach?

**Comment by bluewww** [comment]
Ok thats fine by me.

Do you want me to pu a `rev` tag onto the cv32e40p branch?

**Comment by micprog** [comment]
> Do you want me to put a `rev` tag onto the cv32e40p branch?

Would probably be cleanest, just know that dependency resolution will not work the same as with `version` tags. The benefits are limited to persistence within the repository, and legibility, which I think both is worth it.

**Comment by bluewww** [comment]
Ok I created a tag.
